### PR TITLE
Fixes #3 issue

### LINKED ISSUE #3: Catmull-Rom does not work when two following points have the same coordinates
Hello,

When two following points have the same coordinates the Catmull-Rom algorithm returns a control point with { NaN, NaN } value that makes [path stroke] to stall.

Hope this helps.

### PR DESCRIPTION
**Pull request by letatas**
Fixes Catmull-Rom algorithm not working when two following points have the same coordinates

### CODE REVIEW

File: Curve Interpolation/UIBezierPath+Interpolation.m [hunk 1]
```diff
@@ -36,15 +36,25 @@ +(UIBezierPath *)interpolateCGPointsWithCatmullRom:(NSArray *)pointsAsNSValues c
         float d3 = ccpLength(ccpSub(p3, p2));
         
         CGPoint b1, b2;
-        b1 = ccpMult(p2, powf(d1, 2*alpha));
-        b1 = ccpSub(b1, ccpMult(p0, powf(d2, 2*alpha)));
-        b1 = ccpAdd(b1, ccpMult(p1,(2*powf(d1, 2*alpha) + 3*powf(d1, alpha)*powf(d2, alpha) + powf(d2, 2*alpha))));
-        b1 = ccpMult(b1, 1.0 / (3*powf(d1, alpha)*(powf(d1, alpha)+powf(d2, alpha))));
+        if (d1 == 0) {
```
Comment (jnfisher) [inline Curve Interpolation/UIBezierPath+Interpolation.m:39]: Should this check against a suitably low epsilon or tolerance value instead?
e.g.,
if (d1 < epsilon) {
}

File: Curve Interpolation/UIBezierPath+Interpolation.m [hunk 2]
```diff
@@ -36,15 +36,25 @@ +(UIBezierPath *)interpolateCGPointsWithCatmullRom:(NSArray *)pointsAsNSValues c
...
+            b1 = ccpSub(b1, ccpMult(p0, powf(d2, 2*alpha)));
+            b1 = ccpAdd(b1, ccpMult(p1,(2*powf(d1, 2*alpha) + 3*powf(d1, alpha)*powf(d2, alpha) + powf(d2, 2*alpha))));
+            b1 = ccpMult(b1, 1.0 / (3*powf(d1, alpha)*(powf(d1, alpha)+powf(d2, alpha))));
+        }
         
-        b2 = ccpMult(p1, powf(d3, 2*alpha));
-        b2 = ccpSub(b2, ccpMult(p3, powf(d2, 2*alpha)));
-        b2 = ccpAdd(b2, ccpMult(p2,(2*powf(d3, 2*alpha) + 3*powf(d3, alpha)*powf(d2, alpha) + powf(d2, 2*alpha))));
-        b2 = ccpMult(b2, 1.0 / (3*powf(d3, alpha)*(powf(d3, alpha)+powf(d2, alpha))));
+        if (d3 == 0) {
```
Comment (jnfisher) [inline Curve Interpolation/UIBezierPath+Interpolation.m:49]: Same question as above, 
e.g.,
if (d3 < epsilon) {
}

File: Curve Interpolation/UIBezierPath+Interpolation.m [hunk 3]
```diff
@@ -17,7 +17,7 @@ +(UIBezierPath *)interpolateCGPointsWithCatmullRom:(NSArray *)pointsAsNSValues c
     
     int endIndex = (closed ? [pointsAsNSValues count] : [pointsAsNSValues count]-2);
     NSAssert(alpha >= 0.0 && alpha <= 1.0, @"alpha value is between 0.0 and 1.0, inclusive");
-
+
```
Comment (jnfisher) [inline Curve Interpolation/UIBezierPath+Interpolation.m:20]: Could define epsilon here for instance,

double epsilon = 1.0e-3;

**Comment by jnfisher** [comment]
Thanks for this- looks reasonable aside from my comments.  Generally we should avoid checking floating point numbers with ==.  If we can change that I am happy to merge. :+1:

**Comment by letatas** [comment]
In fact this is just to prevent the div by zero, and the NaN in the coordinates. If the delta is pretty small but not zero the calculation will still be correct. But I agree with you, depending on the float implementation it's safer to use an epsilon.

M.

**Comment by jnfisher** [comment]
Numerical issues can arise with a small enough denominator (results in an overflow or something).  Since this clearly fixes the 0 issue I will merge it.  Thanks!

**Pull request merged**